taichi/ir/expr.cpp

```diff
@@ -127,17 +127,17 @@ Expr Expr::eval() const {
 
 void Expr::operator+=(const Expr &o) {
   if (this->atomic) {
-    current_ast_builder().insert(Stmt::make<FrontendAtomicStmt>(
-        AtomicOpType::add, ptr_if_global(*this), load_if_ptr(o)));
+    (*this) = Expr::make<AtomicOpExpression>(
+        AtomicOpType::add, ptr_if_global(*this), load_if_ptr(o));
   } else {
     (*this) = (*this) + o;
   }
 }
 
 void Expr::operator-=(const Expr &o) {
   if (this->atomic) {
-    current_ast_builder().insert(Stmt::make<FrontendAtomicStmt>(
-        AtomicOpType::add, *this, -load_if_ptr(o)));
+    (*this) = Expr::make<AtomicOpExpression>(
+        AtomicOpType::sub, ptr_if_global(*this), load_if_ptr(o));
```

Comment:
Why `ptr_if_global` appeared? I thought it's already done in L856?

Reply:
I think it's a bug in the old code. See L131 *in the old code*, which has `ptr_if_global`. Also L856 checks for Global**Ptr**Expression, whereas checks for Global**Variable**Expression.

Reply:
Actually, my feeling is `ptr_if_global` should be removed completely in a future PR. We use this wrapper just to allow the syntax like `a += 1` for `a = ti.var(ti.f32, shape())` (0-D tensors). Enforcing the user to write ``a[None] += 1` will make learning easier since the rule is simpler (though more verbose).

Reply:
ah ok, then let's do that in a separate PR..